### PR TITLE
Upgrade Spring Boot & Gradle Wrapper versions across the board

### PR DESCRIPTION
**Pull request by dmikusa**
## Summary

Update everything in one PR.

## Checklist

* [ ] I have viewed, signed, and submitted the Contributor License Agreement.
* [ ] I have linked issue(s) that this PR should close using keywords or the Github UI (See [docs](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue))
* [ ] I have added an integration test, if necessary.
* [ ] I have reviewed the [styleguide](https://github.com/paketo-buildpacks/community/blob/main/STYLEGUIDE.md) for guidance on my code quality.
* [ ] I'm happy with the commit history on this PR (I have rebased/squashed as needed).

### CODE REVIEW

**Comment by anthonydahanne** [comment]
is it stuck because of Spring Boot gradle upgrades?

File: .github/workflows/test-pull-request-java-akka.yml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+jobs:
+  smoke:
+    name: Smoke Tests
+    runs-on: ubuntu-latest
+    steps:
+    - name: Setup Go
+      uses: actions/setup-go@v4
+      with:
+        go-version: 1.18.x
```
Comment (dmikusa) [inline .github/workflows/test-pull-request-java-akka.yml:19]: We should bump these to 1.20

File: .github/workflows/test-pull-request-java-akka.yml [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+    - name: Checkout
+      uses: actions/checkout@v4
+
+    - name: Test Java Samples
+      run: |
+        ./scripts/smoke.sh --suite java/akka/smoke_test \
+                           --builder paketobuildpacks/builder:full \
```
Comment (dmikusa) [inline .github/workflows/test-pull-request-java-akka.yml:27]: Let's drop full. The value we're going to get out of this is questionable, IMHO. Most Java users are on tiny or base.

File: java/akka/smoke_test/akka_test.go [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+					Expect(err).To(MatchError("failed to remove docker container: exit status 1: Container name cannot be empty"))
+				} else {
+					Expect(err).ToNot(HaveOccurred())
+				}
+
+				Expect(docker.Volume.Remove.Execute(occam.CacheVolumeNames(name))).To(Succeed())
+
+				err = docker.Image.Remove.Execute(image.ID)
+				if err != nil {
+					Expect(err.Error()).To(ContainSubstring("failed to remove docker image: exit status 1: Error"))
```
Comment (dmikusa) [inline java/akka/smoke_test/akka_test.go:84]:
```suggestion
					Expect(err).To(MatchError(ContainSubstring("failed to remove docker image: exit status 1: Error")))
```

I'm pretty sure You can nest matchers so you can get substring match on an error.

File: java/akka/smoke_test/akka_test.go [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+					Expect(err).ToNot(HaveOccurred())
+				}
+
+				Expect(os.RemoveAll(source)).To(Succeed())
+			})
+
+			context("app uses akka", func() {
+				it("builds successfully", func() {
+					if strings.HasSuffix(builder, "tiny") {
+						return // this sample requires bash, does not run on tiny
```
Comment (dmikusa) [inline java/akka/smoke_test/akka_test.go:95]: Don't need this any more, just drop tiny from the list in the workflow. Since we're not sharin go workflows.

**Review by dmikusa — COMMENTED** [review]
Thoughts

File: .github/workflows/test-pull-request-java-akka.yml (hunk 1 above)
Comment (anthonydahanne) [reply .github/workflows/test-pull-request-java-akka.yml]: true!

File: .github/workflows/test-pull-request-java-akka.yml (hunk 2 above)
Comment (anthonydahanne) [reply .github/workflows/test-pull-request-java-akka.yml]: fine by me!

File: java/akka/smoke_test/akka_test.go (hunk 4 above)
Comment (anthonydahanne) [reply java/akka/smoke_test/akka_test.go:95]: well, see, I found a way to share 1 single workflow for all Java!

So I'd keep it for now!

**Comment by anthonydahanne** [comment]
Hello @dmikusa 
So this is what I want to generalize for all Java samples (for now just 2)
https://github.com/paketo-buildpacks/samples/actions/runs/6294008765
Continuing refactoring the other samples

**Comment by dmikusa** [comment]
> Hello @dmikusa So this is what I want to generalize for all Java samples (for now just 2) https://github.com/paketo-buildpacks/samples/actions/runs/6294008765 Continuing refactoring the other samples

That's interesting, so instead of using the trigger to execute the workflow based on the path, you have one workflow that runs and then determines what to execute and kicks off other things via a matrix. Did I follow that right?

I don't have strong feelings about the approach so long as we are only testing things that have changed. I do generally have an aversion to lots of inline bash. We have that in a few places and it's a nightmare to maintain. You might want to consider replacing that with an action (i.e. some Go code and unit tests).

If your concern with generalizing stuff is reducing duplication in the workflows, you might also consider a [reusable workflow](https://docs.github.com/en/actions/using-workflows/reusing-workflows#creating-a-reusable-workflow).

**Comment by anthonydahanne** [comment]
> That's interesting, so instead of using the trigger to execute the workflow based on the path, you have one workflow that runs and then determines what to execute and kicks off other things via a matrix. Did I follow that right?

That is correct: there's the first job, named `prepare` that uses git history to find out what are the Java samples that got changed, then it creates a list (as seen in the run summary) of all the Java samples paths that do have a `smoke_test` folder and have changed. Then, a job running `smoke_test` will be created for each element of the list.

>I don't have strong feelings about the approach so long as we are only testing things that have changed. 

Right, this is exactly what it does

> I do generally have an aversion to lots of inline bash. We have that in a few places and it's a nightmare to maintain. You might want to consider replacing that with an action (i.e. some Go code and unit tests).

I understand your concern. I usually also dislike `bash` scripts for anything bigger than 3 lines of code; but...
* I started with `bash` re using some logic I already successfully (as in no maintenance) implemented on another project
* the situation is critical for the `samples` nothing runs for Java; I don't like to use the emergency argument, but rewriting in Go with an image published, etc. well, it will push the fix further away I believe...

> If your concern with generalizing stuff is reducing duplication in the workflows, you might also consider a [reusable workflow](https://docs.github.com/en/actions/using-workflows/reusing-workflows#creating-a-reusable-workflow).

oh, interesting; but my concern was not only about copy/pasting (which re usable workflows in a way solve); but also in the fact of creating a `test-pull-request-XXX.yml` for every and each project; which reusable workflows won't solve unfortunately

**Comment by dmikusa** [comment]
> I understand your concern. I usually also dislike bash scripts for anything bigger than 3 lines of code; but...

Yes, I get the time crunch here. My counter-proposal would be make sure it's using `set -eo pipefail` at the top and very heavy comments in the bash script. Explain what everything is doing.

**Comment by anthonydahanne** [comment]
pretty much all set now @dmikusa .

I'll rebase it on https://github.com/paketo-buildpacks/samples/pull/625 once it's merged and then we're good I believe!

**Comment by dmikusa** [comment]
It's looking good. Only question. I see that the caching host-level caching for Gradle/Maven was removed. Even splitting things up, that's going to help speed up runs. The download time for dependencies is about 1/3 of the build time for Java apps (non-native). Still seems significant. Wondering your thoughts on removing it.

**Comment by anthonydahanne** [comment]
Hello 👋
> I see that the caching host-level caching for Gradle/Maven was removed. Even splitting things up, that's going to help speed up runs.

What caching precisely are you referring to?

In any way, I believe caching for the samples to not be helping.
Sure, we could save some seconds here and there; but this is not what our users are going to experience when they use the paketo buildpacks on their machines: they'll go through long downloads  and potential missing artifacts. As we've ourselves experienced recently some discrepancies between the cached samples and samples builds from scratch, caching can in our case hide environmental issues that we need to notice before our users do.

File: .github/workflows/test-pull-request-java.yml [hunk 5]
```diff
@@ -1,63 +1,87 @@
...
-      run: |
-        #!/usr/bin/env bash
-        set -euo pipefail
-
-        mkdir -p ~/.m2/repository
-        mkdir -p ~/.m2/repository/wrapper/dists
-        mkdir -p ~/.gradle/caches
-        mkdir -p ~/.gradle/wrapper/dists
-        chmod -R 775 ~/.m2
-        chmod -R 775 ~/.gradle
```
Comment (dmikusa) [inline .github/workflows/test-pull-request-java.yml:56]: This part @anthonydahanne

Comment (anthonydahanne) [reply .github/workflows/test-pull-request-java.yml:56]: @dmikusa and I agreed on letting caching go for now; ta favor simplicity.

If things go south, I'll re implement it

**Review by ForestEckhardt — APPROVED** [review]

**Pull request merged**